### PR TITLE
CH-360 Provisoner: Field deployment: Add cluster setup component

### PR DESCRIPTION
**Pull request by Rajnish-TheGreat**
# Problem Statement
- [CH-360 Provisoner: Field deployment: Add cluster setup component](https://jts.seagate.com/browse/CH-360)

# Design
-  K8s install depots for images and check it doesn't for internet access will be done in [CH-342](https://jts.seagate.com/browse/CH-342)
-  Readme file for Ansible Roles will be updated later

# Coding
   Checklist for Author
-  [x] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [x] Unit and System Tests are added
- [ ] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [ ] Testing was performed with RPM

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [ ] Interface change (if any) are documented
- [ ] Side effects on other features (deployment/upgrade)
- [ ] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [x] JIRA number/GitHub Issue added to PR
- [x] PR is self reviewed
- [x] Jira and state/status is updated and JIRA is updated with PR link
- [x] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

# Testing
- Unit Testing
Link: [https://jts.seagate.com/secure/attachment/532454/Unittest_CH-360_PR%2340](https://jts.seagate.com/secure/attachment/532454/Unittest_CH-360_PR%2340)

[WARNING]: Invalid characters were found in group names but not replaced, use -vvvv to see details

PLAY [all] ************************************************************************************************************

TASK [Gathering Facts] ************************************************************************************************
ok: [ssc-vm-g2-rhev4-3412.colo.seagate.com]
ok: [ssc-vm-g2-rhev4-3414.colo.seagate.com]

TASK [common_setup : configure bridging] ******************************************************************************
ok: [ssc-vm-g2-rhev4-3414.colo.seagate.com]
ok: [ssc-vm-g2-rhev4-3412.colo.seagate.com]

TASK [common_setup : Update iptables config for bridging] *************************************************************
ok: [ssc-vm-g2-rhev4-3412.colo.seagate.com]
ok: [ssc-vm-g2-rhev4-3414.colo.seagate.com]

TASK [common_setup : running sysctl] **********************************************************************************
ok: [ssc-vm-g2-rhev4-3412.colo.seagate.com]
ok: [ssc-vm-g2-rhev4-3414.colo.seagate.com]

TASK [common_setup : start docker service] ****************************************************************************
ok: [ssc-vm-g2-rhev4-3412.colo.seagate.com]
ok: [ssc-vm-g2-rhev4-3414.colo.seagate.com]

TASK [common_setup : Changing permissions for docker] *****************************************************************
ok: [ssc-vm-g2-rhev4-3414.colo.seagate.com]
ok: [ssc-vm-g2-rhev4-3412.colo.seagate.com]

TASK [common_setup : Adding docker user] ******************************************************************************
skipping: [ssc-vm-g2-rhev4-3414.colo.seagate.com]
skipping: [ssc-vm-g2-rhev4-3412.colo.seagate.com]

TASK [common_setup : changing cgroupfs driver of docker] **************************************************************
ok: [ssc-vm-g2-rhev4-3414.colo.seagate.com]
ok: [ssc-vm-g2-rhev4-3412.colo.seagate.com]

TASK [common_setup : Adding --cgroup-driver=cgroupfs to conf file] ****************************************************
ok: [ssc-vm-g2-rhev4-3414.colo.seagate.com]
ok: [ssc-vm-g2-rhev4-3412.colo.seagate.com]

TASK [common_setup : Removing if a config file exist] *****************************************************************
ok: [ssc-vm-g2-rhev4-3414.colo.seagate.com]
ok: [ssc-vm-g2-rhev4-3412.colo.seagate.com]

TASK [common_setup : start and enable kubelet] ************************************************************************
ok: [ssc-vm-g2-rhev4-3414.colo.seagate.com]
ok: [ssc-vm-g2-rhev4-3412.colo.seagate.com]

PLAY [master-node] ****************************************************************************************************

TASK [Gathering Facts] ************************************************************************************************
ok: [ssc-vm-g2-rhev4-3414.colo.seagate.com]

TASK [master_setup : checking if kubelet is running or not] ***********************************************************
ok: [ssc-vm-g2-rhev4-3414.colo.seagate.com]

TASK [master_setup : executing kubeadm init command] ******************************************************************
failed: [ssc-vm-g2-rhev4-3414.colo.seagate.com] (item=ssc-vm-g2-rhev4-3414.colo.seagate.com) => {"ansible_loop_var": "item", "changed": true, "cmd": ["kubeadm", "init"], "delta": "0:00:00.453673", "end": "2022-08-22 00:26:09.657532", "item": "ssc-vm-g2-rhev4-3414.colo.seagate.com", "msg": "non-zero return code", "rc": 1, "start": "2022-08-22 00:26:09.203859", "stderr": "error execution phase preflight: [preflight] Some fatal errors occurred:\n\t[ERROR Port-6443]: Port 6443 is in use\n\t[ERROR Port-10259]: Port 10259 is in use\n\t[ERROR Port-10257]: Port 10257 is in use\n\t[ERROR FileAvailable--etc-kubernetes-manifests-kube-apiserver.yaml]: /etc/kubernetes/manifests/kube-apiserver.yaml already exists\n\t[ERROR FileAvailable--etc-kubernetes-manifests-kube-controller-manager.yaml]: /etc/kubernetes/manifests/kube-controller-manager.yaml already exists\n\t[ERROR FileAvailable--etc-kubernetes-manifests-kube-scheduler.yaml]: /etc/kubernetes/manifests/kube-scheduler.yaml already exists\n\t[ERROR FileAvailable--etc-kubernetes-manifests-etcd.yaml]: /etc/kubernetes/manifests/etcd.yaml already exists\n\t[ERROR Port-10250]: Port 10250 is in use\n\t[ERROR Port-2379]: Port 2379 is in use\n\t[ERROR Port-2380]: Port 2380 is in use\n\t[ERROR DirAvailable--var-lib-etcd]: /var/lib/etcd is not empty\n[preflight] If you know what you are doing, you can make a check non-fatal with `--ignore-preflight-errors=...`\nTo see the stack trace of this error execute with --v=5 or higher", "stderr_lines": ["error execution phase preflight: [preflight] Some fatal errors occurred:", "\t[ERROR Port-6443]: Port 6443 is in use", "\t[ERROR Port-10259]: Port 10259 is in use", "\t[ERROR Port-10257]: Port 10257 is in use", "\t[ERROR FileAvailable--etc-kubernetes-manifests-kube-apiserver.yaml]: /etc/kubernetes/manifests/kube-apiserver.yaml already exists", "\t[ERROR FileAvailable--etc-kubernetes-manifests-kube-controller-manager.yaml]: /etc/kubernetes/manifests/kube-controller-manager.yaml already exists", "\t[ERROR FileAvailable--etc-kubernetes-manifests-kube-scheduler.yaml]: /etc/kubernetes/manifests/kube-scheduler.yaml already exists", "\t[ERROR FileAvailable--etc-kubernetes-manifests-etcd.yaml]: /etc/kubernetes/manifests/etcd.yaml already exists", "\t[ERROR Port-10250]: Port 10250 is in use", "\t[ERROR Port-2379]: Port 2379 is in use", "\t[ERROR Port-2380]: Port 2380 is in use", "\t[ERROR DirAvailable--var-lib-etcd]: /var/lib/etcd is not empty", "[preflight] If you know what you are doing, you can make a check non-fatal with `--ignore-preflight-errors=...`", "To see the stack trace of this error execute with --v=5 or higher"], "stdout": "[init] Using Kubernetes version: v1.24.4\n[preflight] Running pre-flight checks", "stdout_lines": ["[init] Using Kubernetes version: v1.24.4", "[preflight] Running pre-flight checks"]}
...ignoring

TASK [master_setup : creating .kube folder] ***************************************************************************
ok: [ssc-vm-g2-rhev4-3414.colo.seagate.com]

TASK [master_setup : copy config file to .kube folder] ****************************************************************
skipping: [ssc-vm-g2-rhev4-3414.colo.seagate.com]

TASK [master_setup : changing permission of file] *********************************************************************
skipping: [ssc-vm-g2-rhev4-3414.colo.seagate.com]

TASK [master_setup : saving the token] ********************************************************************************
skipping: [ssc-vm-g2-rhev4-3414.colo.seagate.com]

TASK [master_setup : fetch the token command] *************************************************************************
skipping: [ssc-vm-g2-rhev4-3414.colo.seagate.com]

TASK [master_setup : installing weavenet for vxlan] *******************************************************************
skipping: [ssc-vm-g2-rhev4-3414.colo.seagate.com]

PLAY [worker-node] ****************************************************************************************************

TASK [Gathering Facts] ************************************************************************************************
ok: [ssc-vm-g2-rhev4-3412.colo.seagate.com]

TASK [worker_setup : copy the join command file] **********************************************************************
ok: [ssc-vm-g2-rhev4-3412.colo.seagate.com]

TASK [worker_setup : running kubeadm join command] ********************************************************************
skipping: [ssc-vm-g2-rhev4-3412.colo.seagate.com]

PLAY [master-node] ****************************************************************************************************

TASK [Gathering Facts] ************************************************************************************************
ok: [ssc-vm-g2-rhev4-3414.colo.seagate.com]

TASK [final_setup : Assigning roles to worker nodes] ******************************************************************
fatal: [ssc-vm-g2-rhev4-3414.colo.seagate.com]: FAILED! => {"msg": "Invalid data passed to 'loop', it requires a list, got this instead: groups['worker-node']. Hint: If you passed a list/dict of just one element, try adding wantlist=True to your lookup invocation or use q/query instead of lookup."}
...ignoring

TASK [final_setup : Untaint master node] ******************************************************************************
fatal: [ssc-vm-g2-rhev4-3414.colo.seagate.com]: FAILED! => {"changed": true, "cmd": "kubectl taint nodes --all node-role.kubernetes.io/master-\nkubectl taint nodes --all node-role.kubernetes.io/control-plane- ", "delta": "0:00:00.258007", "end": "2022-08-22 00:26:19.622371", "msg": "non-zero return code", "rc": 1, "start": "2022-08-22 00:26:19.364364", "stderr": "taint \"node-role.kubernetes.io/master\" not found\ntaint \"node-role.kubernetes.io/master\" not found\ntaint \"node-role.kubernetes.io/control-plane\" not found\ntaint \"node-role.kubernetes.io/control-plane\" not found", "stderr_lines": ["taint \"node-role.kubernetes.io/master\" not found", "taint \"node-role.kubernetes.io/master\" not found", "taint \"node-role.kubernetes.io/control-plane\" not found", "taint \"node-role.kubernetes.io/control-plane\" not found"], "stdout": "", "stdout_lines": []}
...ignoring

PLAY RECAP ************************************************************************************************************
ssc-vm-g2-rhev4-3412.colo.seagate.com : ok=12   changed=0    unreachable=0    failed=0    skipped=2    rescued=0    ignored=0
ssc-vm-g2-rhev4-3414.colo.seagate.com : ok=17   changed=2    unreachable=0    failed=0    skipped=6    rescued=0    ignored=3

[root@ssc-vm-g2-rhev4-3414 ~]# kubectl get nodes
NAME                                    STATUS   ROLES           AGE    VERSION
ssc-vm-g2-rhev4-3412.colo.seagate.com   Ready    worker          4d4h   v1.24.4
ssc-vm-g2-rhev4-3414.colo.seagate.com   Ready    control-plane   4d5h   v1.24.4

[root@ssc-vm-g2-rhev4-3413 k8s]# py.test test_cluster_setup.py -vvvv
===================================================== test session starts =====================================================
platform linux -- Python 3.6.8, pytest-7.0.1, pluggy-1.0.0 -- /usr/bin/python3
cachedir: .pytest_cache
rootdir: /root/halo-rajnish/cortx-halo, configfile: pytest.ini
collected 1 item                                                                                                              

test_cluster_setup.py::test_setup PASSED                                                                                [100%]

===================================================== 1 passed in 51.54s ======================================================

-----
[View rendered provisioner/components/k8s/setup_playbook/common_setup/README.md](https://github.com/Rajnish-TheGreat/cortx-halo/blob/main/provisioner/components/k8s/setup_playbook/common_setup/README.md)
[View rendered provisioner/components/k8s/setup_playbook/final_setup/README.md](https://github.com/Rajnish-TheGreat/cortx-halo/blob/main/provisioner/components/k8s/setup_playbook/final_setup/README.md)
[View rendered provisioner/components/k8s/setup_playbook/master_setup/README.md](https://github.com/Rajnish-TheGreat/cortx-halo/blob/main/provisioner/components/k8s/setup_playbook/master_setup/README.md)

### CODE REVIEW

**Comment by Rajnish-TheGreat** [comment]
Readme files for Ansible Roles will be updated later

**Review by shaileshsvc — APPROVED** [review]

File: provisioner/components/k8s/setup_playbook/common_setup/.travis.yml [hunk 1]
```diff
@@ -0,0 +1,29 @@
+---
+language: python
+python: "2.7"
```
Comment (ajaysrivas) [inline provisioner/components/k8s/setup_playbook/common_setup/.travis.yml:3]: Is it generated a file ? python version should be > 3.6.

Comment (Rajnish-TheGreat) [reply provisioner/components/k8s/setup_playbook/common_setup/.travis.yml:3]: Yes it's a generated file that gets generated when we create an ansible role

File: provisioner/components/k8s/setup_playbook/final_setup/.travis.yml [hunk 2]
```diff
@@ -0,0 +1,29 @@
+---
+language: python
+python: "2.7"
```
Comment (mukhtar-inamdar) [inline provisioner/components/k8s/setup_playbook/final_setup/.travis.yml:3]: are we using python 2.7 or 3.6?

File: provisioner/components/k8s/setup_playbook/worker_setup/.travis.yml [hunk 3]
```diff
@@ -0,0 +1,29 @@
+---
+language: python
+python: "2.7"
```
Comment (mukhtar-inamdar) [inline provisioner/components/k8s/setup_playbook/worker_setup/.travis.yml:3]: are we using python 2.7 or python 3.6?

File: provisioner/components/k8s/test_cluster_setup.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+from cluster_setup import setup
+from cluster_validate import validate
+
+
+def test_setup():
+    setup()
+    x = validate()
+    assert x!=0, "Cluster setup failed"
+
+
```
Comment (mukhtar-inamdar) [inline provisioner/components/k8s/test_cluster_setup.py:29]: why so many lines here?
also, check 2 empty lines I can see in between functions.
I think standard is set to 1 line, please check

File: provisioner/components/k8s/validate_playbook/cluster_validate_playbook.yml [hunk 5]
```diff
@@ -0,0 +1,4 @@
+- hosts: master-node
+  become: true
+  roles:
+    - role: validate_cluster
```
Comment (mukhtar-inamdar) [inline provisioner/components/k8s/validate_playbook/cluster_validate_playbook.yml:4]: are some files showing the red sign as no newline at the end?
so please add a new line at the end wherever such line can be seen.

File: provisioner/components/k8s/validate_playbook/validate_cluster/.travis.yml [hunk 6]
```diff
@@ -0,0 +1,29 @@
+---
+language: python
+python: "2.7"
```
Comment (mukhtar-inamdar) [inline provisioner/components/k8s/validate_playbook/validate_cluster/.travis.yml:3]: 2.7 or 3.6?

**Review by mukhtar-inamdar — APPROVED** [review]
some minor comments and questions address it then you can continue

File: provisioner/components/k8s/setup_playbook/final_setup/.travis.yml (hunk 2 above)
Comment (Rajnish-TheGreat) [reply provisioner/components/k8s/setup_playbook/final_setup/.travis.yml:3]: We are using python 3.6 only but it's a generated file that gets generated when we create an ansible role

File: provisioner/components/k8s/setup_playbook/worker_setup/.travis.yml (hunk 3 above)
Comment (Rajnish-TheGreat) [reply provisioner/components/k8s/setup_playbook/worker_setup/.travis.yml:3]: We are using python 3.6 only but it's a generated file that gets generated when we create an ansible role

File: provisioner/components/k8s/validate_playbook/validate_cluster/.travis.yml (hunk 6 above)
Comment (Rajnish-TheGreat) [reply provisioner/components/k8s/validate_playbook/validate_cluster/.travis.yml:3]: We are using python 3.6 only but it's a generated file that gets generated when we create an ansible role

File: provisioner/components/k8s/test_cluster_setup.py (hunk 4 above)
Comment (Rajnish-TheGreat) [reply provisioner/components/k8s/test_cluster_setup.py]: Fixed it

File: provisioner/components/k8s/validate_playbook/cluster_validate_playbook.yml (hunk 5 above)
Comment (Rajnish-TheGreat) [reply provisioner/components/k8s/validate_playbook/cluster_validate_playbook.yml]: Fixed it

File: provisioner/components/k8s/setup_playbook/common_setup/.travis.yml (hunk 1 above)
Comment (Rajnish-TheGreat) [reply provisioner/components/k8s/setup_playbook/common_setup/.travis.yml]: Updated with python 3.6

File: provisioner/components/k8s/setup_playbook/final_setup/.travis.yml (hunk 2 above)
Comment (Rajnish-TheGreat) [reply provisioner/components/k8s/setup_playbook/final_setup/.travis.yml]: Updated with python 3.6

File: provisioner/components/k8s/setup_playbook/worker_setup/.travis.yml (hunk 3 above)
Comment (Rajnish-TheGreat) [reply provisioner/components/k8s/setup_playbook/worker_setup/.travis.yml]: Updated with python 3.6

File: provisioner/components/k8s/validate_playbook/validate_cluster/.travis.yml (hunk 6 above)
Comment (Rajnish-TheGreat) [reply provisioner/components/k8s/validate_playbook/validate_cluster/.travis.yml]: Updated with python 3.6

**Review by shaileshsvc — APPROVED** [review]

File: provisioner/components/k8s/cluster_setup.py [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+# You should have received a copy of the GNU Affero General Public License
+# along with this program. If not, see <https://www.gnu.org/licenses/>.
+# For any questions about this software or licensing, please email
+# opensource@seagate.com or cortx-questions@seagate.com.
+
+import ansible_runner
+import logging
+
+inventory_path = "/opt/halo/install_depot/cortx-halo/provisioner/components/k8s/setup_playbook/inventory.yaml"
```
Comment (ajaysrivas) [inline provisioner/components/k8s/cluster_setup.py:23]: These variables should go to const.py at root directory.
HALO_BASE_PATH = "/opt/seagate/halo/"
PROVISIONER_PATH = HALO_BASE_PATH + 'install_depot/cortx-halo/provisioner/'
INVENTORY_PATH and so one.

File: provisioner/components/k8s/cluster_setup.py [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+    def run (playbook, ini):
+        rc = ansible_runner.run(playbook = playbook_path, inventory = inventory_path)
+        return rc
+
+
+def setup():
+    try:
+        return(Runner.run(playbook_path, inventory_path))
+    except Exception as e:
+        logging.exception(f'{e}')
```
Comment (ajaysrivas) [inline provisioner/components/k8s/cluster_setup.py:38]: Look at logging conf.

**Review by ajaysrivas — CHANGES_REQUESTED** [review]

File: provisioner/components/k8s/setup_playbook/common_setup/.travis.yml (hunk 1 above)
Comment (ajaysrivas) [reply provisioner/components/k8s/setup_playbook/common_setup/.travis.yml]: Please check if code works with python3.6 by removing python 2.x from the system.

File: provisioner/components/k8s/cluster_setup.py (hunk 7 above)
Comment (Rajnish-TheGreat) [reply provisioner/components/k8s/cluster_setup.py]: Moved variables to const.py

File: provisioner/components/k8s/cluster_setup.py (hunk 8 above)
Comment (Rajnish-TheGreat) [reply provisioner/components/k8s/cluster_setup.py]: Added log store in file

File: provisioner/components/k8s/cluster_validate.py [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+# along with this program. If not, see <https://www.gnu.org/licenses/>.
+# For any questions about this software or licensing, please email
+# opensource@seagate.com or cortx-questions@seagate.com.
+
+import ansible_runner
+import logging
+from provisioner.const import PATH
+
+
+logging.basicConfig(filename=PATH.LOG_FILE,
```
Comment (ajaysrivas) [inline provisioner/components/k8s/cluster_validate.py:25]: This is duplicate code. Put logging configuration in different file and import logger from there in different modules.

Comment (Rajnish-TheGreat) [reply provisioner/components/k8s/cluster_validate.py]: Removed Logging for now. Will add it later

File: provisioner/components/k8s/cluster_setup.py (hunk 8 above)
Comment (Rajnish-TheGreat) [reply provisioner/components/k8s/cluster_setup.py]: Removed Logging for now. Will add it later

File: provisioner/const.py [hunk 10]
```diff
@@ -35,3 +35,14 @@ class SOFTWARE(Enum):
...
+    HALO_BASE_PATH = '/opt/seagate/halo/install_depot/'
+    PROVISIONER_PATH = HALO_BASE_PATH + 'cortx-halo/provisioner/'
+    ANSIBLE_PATH = PROVISIONER_PATH + 'components/ansible/'
+    K8S_PATH = PROVISIONER_PATH + 'components/k8s/'
+    MONITOR_PATH = PROVISIONER_PATH + 'components/monitor/'
+    NODECLI_PATH = PROVISIONER_PATH + 'components/nodecli/'
+    ANSIBLE_INVENTORY_PATH = K8S_PATH + 'inventory.yaml'
+    CLUSTER_SETUP_PLAYBOOK_PATH =  K8S_PATH + 'setup_playbook/cluster_setup_playbook.yml'
+    CLUSTER_VALIDATE_PLAYBOOK_PATH = K8S_PATH + 'validate_playbook/cluster_validate_playbook.yml'
+    LOG_FILE = PROVISIONER_PATH + 'provisioner.log'
```
Comment (ajaysrivas) [inline provisioner/const.py:48]: This should be LOG_FILE = PROVISIONER_PATH + 'logs/provisioner.log'

Comment (Rajnish-TheGreat) [reply provisioner/const.py]: Update Path

File: provisioner/components/k8s/setup_playbook/common_setup/.travis.yml (hunk 1 above)
Comment (Rajnish-TheGreat) [reply provisioner/components/k8s/setup_playbook/common_setup/.travis.yml]: Yes working

**Review by ajaysrivas — APPROVED** [review]

**Pull request merged**